### PR TITLE
purge ephemeral messages periodically in the background

### PR DESCRIPTION
**Pull request by joshblum**
depends on https://github.com/keybase/client/pull/11388

This is part 2 of purging ephemeral messages. A timer is added to the background conversation loader to periodically purge any conversations that have ephemeral messages that have exploded

### CODE REVIEW

File: go/chat/boxer.go [hunk 1]
```diff
@@ -749,6 +749,7 @@ func (b *Boxer) unversionHeaderMBV2(ctx context.Context, headerVersioned chat1.H
 			OutboxID:          hp.OutboxID,
 			OutboxInfo:        hp.OutboxInfo,
 			KbfsCryptKeysUsed: hp.KbfsCryptKeysUsed,
+			EphemeralMetadata: hp.EphemeralMetadata, // TODO only for v3
```
Comment (joshblum) [inline go/chat/boxer.go]: this is a temporary measure to get tests running but should be changed in https://github.com/keybase/client/pull/11404/files

File: go/chat/storage/storage_ephemeral_purge.go [hunk 2]
```diff
@@ -58,11 +82,6 @@ func (s *Storage) EphemeralPurge(ctx context.Context, convID chat1.ConversationI
 	if err != nil {
 		return nil, err
 	}
-	// End of the line
-	if newPurgeInfo.MinUnexplodedID == maxMsgID {
```
Comment (joshblum) [inline go/chat/storage/storage_ephemeral_purge.go]: we don't want to delete here since we'll lose our place in the chat and have to sweep over the entire chat on the next ephemeral message that explodes

File: go/chat/types/types.go [hunk 3]
```diff
@@ -81,41 +81,67 @@ type Inbox struct {
...
+		return fmt.Sprintf("[EphemeralPurgeBackgroundJob convID: %s purgeInfo: %v]", j.ConvID, j.PurgeInfo)
+	default:
+		return fmt.Sprintf("[convID: %v]", j.ConvID)
+	}
 }
 
-func NewConvLoaderJob(convID chat1.ConversationID, pagination *chat1.Pagination, priority ConvLoaderPriority,
-	postLoadHook func(context.Context, chat1.ThreadView, ConvLoaderJob)) ConvLoaderJob {
-	return ConvLoaderJob{
+func NewConvLoaderJob(convID chat1.ConversationID, pagination *chat1.Pagination, priority BgLoaderPriority,
```
Comment (joshblum) [inline go/chat/types/types.go]: This seems a little messy -- is there a cleaner way?

File: go/chat/convloader.go [hunk 4]
```diff
@@ -20,13 +20,15 @@ import (
...
+	bgLoaderMaxAttempts   = 10
+	bgLoaderInitDelay     = 100 * time.Millisecond
+	bgLoaderErrDelay      = 300 * time.Millisecond
+	bgLoaderPurgeTick     = 5 * time.Minute
+	bgLoaderPurgeTickTest = 500 * time.Millisecond
 )
 
 type clTask struct {
-	job           types.ConvLoaderJob
+	job           types.BgLoaderJob
```
Comment (mmaxim) [inline go/chat/convloader.go]: Let's not rename this

File: go/chat/convloader.go [hunk 5]
```diff
@@ -107,22 +109,25 @@ type BackgroundConvLoader struct {
...
+	ephemeralTick time.Duration
+	storage       *storage.Storage
 
 	activeLoadCtx      context.Context
 	activeLoadCancelFn context.CancelFunc
 	suspendCount       int
 
 	// for testing, make this and can check conv load successes
-	loads                 chan chat1.ConversationID
+	loadTestCh            chan chat1.ConversationID
```
Comment (mmaxim) [inline go/chat/convloader.go]: Can we just have one channel here?

File: go/chat/types/types.go [hunk 6]
```diff
@@ -107,7 +107,7 @@ func (j ConvLoaderJob) HigherPriorityThan(j2 ConvLoaderJob) bool {
 }
 
 func (j ConvLoaderJob) String() string {
-	return fmt.Sprintf("[convID: %s pagination: %s]", j.ConvID, j.Pagination)
+	return fmt.Sprintf("[convID: %s pagination: %v]", j.ConvID, j.Pagination)
```
Comment (joshblum) [inline go/chat/types/types.go]: i got a `go vet` error here..
```
go vet...................................................................Failed
hookid: go-vet

go/chat/types/types.go:110: arg j.Pagination for printf verb %s of wrong type: *github.com/keybase/client/go/protocol/chat1.Pagination
exit status 1
```

Comment (mmaxim) [reply go/chat/types/types.go]: Hmm I think you should change this back, are you merged up against master? `chat1.Pagination` has a `String` function on it.

File: go/chat/convloader.go [hunk 7]
```diff
@@ -426,6 +430,7 @@ func (b *BackgroundConvLoader) load(ictx context.Context, task clTask, uid grego
 		b.Unlock()
 	}()
 
+	job := task.job
```
Comment (mmaxim) [inline go/chat/convloader.go]: Why did you move this?

File: go/chat/convloader.go [hunk 8]
```diff
@@ -113,16 +113,17 @@ type BackgroundConvLoader struct {
 	activeLoadCtx      context.Context
 	activeLoadCancelFn context.CancelFunc
 	suspendCount       int
+	storage            *storage.Storage
 
 	// for testing, make this and can check conv load successes
-	loads                 chan chat1.ConversationID
+	loadTestCh            chan chat1.ConversationID
```
Comment (mmaxim) [inline go/chat/convloader.go]: Can you just rename this back to original name?

File: go/chat/convloader.go [hunk 9]
```diff
@@ -448,15 +453,46 @@ func (b *BackgroundConvLoader) load(ictx context.Context, task clTask, uid grego
...
+	if b.loadTestCh != nil {
+		b.Debug(ctx, "load: putting convID %s on loadTestCh", job.ConvID)
+		b.loadTestCh <- job.ConvID
 	}
+
 	return nil
 }
 
+// Called periodically by the fastChatChecks ticker in service/main
+func (b *BackgroundConvLoader) QueueEphemeralPurges(ctx context.Context) {
```
Comment (mmaxim) [inline go/chat/convloader.go]: I'd prefer for this to not be a function on the conv loader. Maybe in `utils` somewhere? Or just next to the loop that is driving it?

File: go/chat/convloader.go [hunk 10]
```diff
@@ -113,16 +113,17 @@ type BackgroundConvLoader struct {
...
-	loads                 chan chat1.ConversationID
+	loadTestCh            chan chat1.ConversationID
 	testingNameInfoSource types.NameInfoSource
 	appStateCh            chan struct{}
 }
 
 var _ types.ConvLoader = (*BackgroundConvLoader)(nil)
 
-func NewBackgroundConvLoader(g *globals.Context) *BackgroundConvLoader {
+func NewBackgroundConvLoader(g *globals.Context, storage *storage.Storage) *BackgroundConvLoader {
```
Comment (mmaxim) [inline go/chat/convloader.go]: I'd like it if this guy didn't require a storage, see my comment below about the location of that `Queue` function.

File: go/chat/convloader.go (hunk 7 above)
Comment (joshblum) [reply go/chat/convloader.go]: just leftover from the old diff, will revert
